src/Components/WebAssembly/WebAssembly/src/Services/LazyAssemblyLoader.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+using System.Runtime.InteropServices;
+using System.Runtime.Loader;
+using Microsoft.JSInterop.WebAssembly;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.WebAssembly.Services
+{
+    public class LazyAssemblyLoader
+    {
+        internal const string GetDynamicAssemblies = "window.Blazor._internal.getLazyAssemblies";
+        internal const string ReadDynamicAssemblies = "window.Blazor._internal.readLazyAssemblies";
+
+        private List<string> _loadedAssemblyCache = new List<string>();
+
+        private readonly IServiceProvider _provider;
+
+        public LazyAssemblyLoader(IServiceProvider provider)
+        {
+            _provider = provider;
+        }
+
+        public async Task<IEnumerable<Assembly>> LoadAssembliesAsync(IEnumerable<string> assembliesToLoad)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Browser))
+            {
+                return await LoadAssembliesInClientAsync(assembliesToLoad);
+            }
+
+            return await LoadAssembliesInServerAsync(assembliesToLoad);
+        }
+
+        private Task<IEnumerable<Assembly>> LoadAssembliesInServerAsync(IEnumerable<string> assembliesToLoad)
+        {
+            var assemblies = AppDomain.CurrentDomain.GetAssemblies().Where(assembly =>
+                assembliesToLoad.Contains(assembly.GetName().Name + ".dll"));
+
+            if (assemblies.Count() != assembliesToLoad.Count())
+            {
+                var unloadedAssemblies = assembliesToLoad.Except(assemblies.Select(assembly => assembly.GetName().Name + ".dll"));
+                throw new FileNotFoundException($"Unable to find the following assemblies: {string.Join(",", unloadedAssemblies)}");
```

Comment:
```suggestion
                var notFoundAssemblies = assembliesToLoad.Except(assemblies.Select(assembly => assembly.GetName().Name + ".dll"));
                throw new InvalidOperationException($"Unable to find the following assemblies: {string.Join(",", notFoundAssemblies). Make sure that the appplication is referencing the assemblies and that they are present in the output folder.}");
```

Reply:
I threw a `FileNotFound` exception here to mimic what the runtime would do if you tried to reference an API for an unloaded assembly. I guess `InvalidOperation` does make more sense in our context though.